### PR TITLE
new receiver `prometheusremotewritereceiver`

### PR DESCRIPTION
**Pull request by adamkubon**
**Description:** New prometheusremotewrite receiver.

**Link to tracking Issue:** #10358

**Testing:** New receiver can consume the prometheus remote write protocol

**Documentation:** Markdown README with config description added

### CODE REVIEW

File: receiver/prometheusremotewritereceiver/receiver.go
Comment (kovrus) [inline receiver/prometheusremotewritereceiver/receiver.go:112]: Move the regex patter compilation outside the method scope.

File: receiver/prometheusremotewritereceiver/receiver.go
Comment (kovrus) [inline receiver/prometheusremotewritereceiver/receiver.go:206]: Return an instance of `error` and handle it properly by the invocation logic.

File: receiver/prometheusremotewritereceiver/receiver.go
Comment (kovrus) [inline receiver/prometheusremotewritereceiver/receiver.go:116]: Regarding the conversion, have you seen https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/translator/prometheusremotewrite?
Maybe we can extract the `prompb.TimeSeries => pmetric.Metrics` conversion logic from `ServeHTTP` and add it to the package I mentioned above with proper test coverage, etc.?

**Review by kovrus — CHANGES_REQUESTED** [review]

File: receiver/prometheusremotewritereceiver/receiver.go
Comment (Aneurysm9) [inline receiver/prometheusremotewritereceiver/receiver.go:110]: This should happen at the top of the method so it captures the decoding time.  Also, use the context from the incoming request:

```suggestion
	ctx := rec.obsrecv.StartMetricsOp(r.Context())
```

File: receiver/prometheusremotewritereceiver/receiver.go
Comment (Aneurysm9) [inline receiver/prometheusremotewritereceiver/receiver.go:113]: Should this be configurable?

File: receiver/prometheusremotewritereceiver/receiver.go
Comment (Aneurysm9) [reply receiver/prometheusremotewritereceiver/receiver.go:116]: This is the way.

File: receiver/prometheusremotewritereceiver/receiver.go
Comment (Aneurysm9) [inline receiver/prometheusremotewritereceiver/receiver.go:178]:
```suggestion
	pm.MoveAndAppendTo(pms.ResourceMetrics().AppendEmpty().ScopeMetrics().AppendEmpty().Metrics())
```

Not sure that creating a new `ResourceMetrics` for every timeseries is appropriate, and there will need to be work done to ensure that the resource and scope attributes are properly extracted, but if this is the way it'll be done best to avoid all the copying.  Better yet, do this all up at the top of this loop:

```go
pm := pms.ResourceMetrics().AppendEmpty().ScopeMetrics().AppendEmpty().Metrics().AppendEmpty()
```

and then fill it in from there.

File: receiver/prometheusremotewritereceiver/config_test.go
Comment (adamkubon) [reply receiver/prometheusremotewritereceiver/config_test.go:30]: done

File: receiver/prometheusremotewritereceiver/config.go
Comment (adamkubon) [reply receiver/prometheusremotewritereceiver/config.go:25]: done

File: receiver/prometheusremotewritereceiver/README.md
Comment (adamkubon) [reply receiver/prometheusremotewritereceiver/README.md:23]: done

File: receiver/prometheusremotewritereceiver/README.md
Comment (adamkubon) [reply receiver/prometheusremotewritereceiver/README.md:2]: done

File: receiver/prometheusremotewritereceiver/receiver.go
Comment (adamkubon) [reply receiver/prometheusremotewritereceiver/receiver.go]: done

File: receiver/prometheusremotewritereceiver/receiver.go
Comment (adamkubon) [reply receiver/prometheusremotewritereceiver/receiver.go]: done

**Comment by jpkrohling** [comment]
@norbertwnuk, could you also please review this one? This cannot be merged until we have at least one sponsor for this component. Besides that, I would like to ask what would be your availability in the future to maintain this component.

**Review by jpkrohling — CHANGES_REQUESTED** [review]
Blocking this from being merged until it has a sponsor.

File: receiver/prometheusremotewritereceiver/receiver.go
Comment (kovrus) [reply receiver/prometheusremotewritereceiver/receiver.go:117]: @adamkubon maybe let's even do it as a separate PR first? Also, [pkg/translator/prometheusremotewrite](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/translator/prometheusremotewrite ) contains `pmetric.Metrics  =>  prompb.TimeSeries` conversion implemented, maybe we can add some "roundtrip" tests after `prompb.TimeSeries => pmetric.Metrics` conversion is implemented to verify conversion logic is consistent. WDYT?

File: receiver/prometheusremotewritereceiver/receiver.go
Comment (adamkubon) [reply receiver/prometheusremotewritereceiver/receiver.go]: parameter added

File: receiver/prometheusremotewritereceiver/receiver.go
Comment (adamkubon) [reply receiver/prometheusremotewritereceiver/receiver.go]: done

**Comment by jpkrohling** [comment]
@adamkubon, @norbertwnuk, are you still interested in this one?

**Comment by adamkubon** [comment]
> 

Yes definitely We are, sorry for late response due to PTO

**Comment by kovrus** [comment]
Hi @adamkubon @norbert, I would like to join you with the work on the remote-write receiver and further share the code ownership for this component.

I have no comments regarding the component itself, but we would have to develop some strategy for Prometheus -> OTLP [metrics mapping.](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/14904) We would have to think about using metadata (probably the receiver will become then stateful). If you look at [`prometheusreceiver` receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver/internal) component, it already contains the mapping logic that we need and uses metadata storage. Maybe we can take a look at it and think whether we can reuse that code, maybe make it public if it makes sense, revise this https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/prometheusreceiver/DESIGN.md#prometheus-metric-to-opentelemetry-metric-proto-mapping documentation, etc. I am happy to discuss it further asynchronously or jump on a call.

**Comment by adamkubon** [comment]
Hi @kovrus we appreciate any help with development the remote-write receiver. Unfortunatelly we don't have so much time this year to spend on it. We think about engage some effort in Q1/2023. Please clarify your plans - what about this pr and the base pr ( #14904). Should we going to merge it in more/less current shape or split the work on it in some how?

**Comment by kovrus** [comment]
> Hi @kovrus we appreciate any help with development the remote-write receiver. Unfortunatelly we don't have so much time this year to spend on it. We think about engage some effort in Q1/2023. Please clarify your plans - what about this pr and the base pr ( #14904). Should we going to merge it in more/less current shape or split the work on it in some how?

Then I can take over this work. First, we would have to find a sponsor for this component, otherwise, nothing can be merged. I'll try to do that. Regarding the current work, #14752 looks good we can update it once the mapping part is done. For #14904, I guess, we would have to do some research on how to implement the mapping, see my previous comment and get input from other Prom components owners. I am going to start looking into the mapping part now, checking how existent components can be reused, and will try to get some input from the community.

**Comment by felix** [comment]
Hi, what is involved in becoming a sponsor?

We at MYOB use remotewrite -> OTLP heavily with our own custom receiver (amongst several others) and may consider sponsorship.

**Comment by kovrus** [comment]
> Hi, what is involved in becoming a sponsor?
> 
> We at MYOB use remotewrite -> OTLP heavily with our own custom receiver (amongst several others) and may consider sponsorship.

Sponsorship is about finding a maintainer or approver who will act as official reviewer and one of code owners for this component, see https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components

**Comment by aunshc** [comment]
@kovrus any luck with finding a sponsor so far?

**Comment by kovrus** [comment]
> @kovrus any luck with finding a sponsor so far?

Yes, see https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14751. I've been doing some research and now working on the design doc for this component.

**Comment by kovrus** [comment]
> Is there any update of this PR? We need this feature. +1
> 
> And seems it's lots of changes in main, and this PR cannot run with the latest main branch, can you update to latest version so we can build a custom binary?

I think that @atoulme mentioned that there was someone who wanted to contribute probably something close to the "best effort" implementation from this [design doc](https://docs.google.com/document/d/1FFDQzs8Y8v5YO1go3HeWQITxB9y3mCc8I-p53SfGBeU/edit?usp=sharing).

**Comment by kelseyma** [comment]
Hi all! Was wondering if there are any updates on this?

**Comment by futurist** [comment]
We need it, please don't close it.

**Comment by jpkrohling** [comment]
Is anyone interested in picking up this one? I'm closing for now, but if you want to continue working on this, I'll reopen it.

**Comment by ernado** [comment]
Hey @jpkrohling, what is the process and requirements for becoming a sponsor?

**Comment by atoulme** [comment]
Approvers and maintainers are sponsors on this repository: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components

**Pull request closed without merging**
